### PR TITLE
Refs #37068 -- Avoided table rebuild on SQLite for altering NOT NULL.

### PR DESCRIPTION
**Pull request by adamchainz**
#### Trac ticket number

ticket-37068

#### Branch description

Deal with "part two" of the ticket, avoiding table rebuilds on SQLite 3.53.1+ when only a field's nullability changes.

I used the same `pysqlite3` strategy as in #21181 (part one) to test on SQLite 3.53.1.

#### AI Assistance Disclosure (REQUIRED)

- [ ] **No AI tools were used** in preparing this PR.
- [x] **If AI tools were used**, I have disclosed which ones, and fully reviewed and verified their output.

I used a bit of Claude.

#### Checklist
- [x] This PR follows the [contribution guidelines](https://docs.djangoproject.com/en/stable/internals/contributing/writing-code/submitting-patches/).
- [x] This PR **does not** disclose a security vulnerability (see [vulnerability reporting](https://docs.djangoproject.com/en/stable/internals/security/)).
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period (see [guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines)).
- [x] I have not requested, and will not request, an automated AI review for this PR. 
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/db/backends/sqlite3/schema.py [hunk 1]
```diff
@@ -383,6 +383,30 @@ def _alter_field(
...
+            and old_field.column == new_field.column
+            and old_type == new_type
+            and old_field.null != new_field.null
+            and old_db_params.get("check") == new_db_params.get("check")
+            and old_db_params.get("collation") == new_db_params.get("collation")
+            and old_field.unique == new_field.unique
+            and old_field.db_index == new_field.db_index
+            and not (old_field.remote_field and old_field.db_constraint)
+            and not (new_field.remote_field and new_field.db_constraint)
+        ):
```
[the remark this answers is not in the document]
Comment (adamchainz) [reply django/db/backends/sqlite3/schema.py:406]: Thank you so much for thoroughly testing! Yes, those tests were broken, I've added the guard now.

> CI likely doesn't catch this because its bundled SQLite is older than 3.53.

Yeah, that'd be it. Django tests with the built-in SQLite only. I have started the process to upgrade SQLite in CPython in https://github.com/python/cpython/issues/149029, which will hopefully ship with Python 3.15.


> Maybe also add a regression test for `IntegerField(null=True) -> IntegerField(null=False, default=7)` that checks `_remake_table()` is used and that an existing NULL row becomes `7`.


I thought about adding exhaustive tests, but that seems like it would be quite verbose. I'm not sure it's worth it…

File: django/db/backends/sqlite3/schema.py [hunk 2]
```diff
@@ -383,6 +383,30 @@ def _alter_field(
                     model._meta.db_table, old_field, new_field, new_type
                 )
             )
+        # On SQLite 3.53.0+, use newly supported "ALTER TABLE ... ALTER COLUMN
+        # ... {SET|DROP} NOT NULL" if only nullability has changed.
+        if (
+            self.connection.Database.sqlite_version_info >= (3, 53)
```
[the remark this answers is not in the document]
Comment (adamchainz) [reply django/db/backends/sqlite3/schema.py]: Done!

Mariusz had the same feedback over on the sister PR to this one: https://github.com/django/django/pull/21181#discussion_r3165713855

Great minds think alike!

**Comment by adamchainz** [comment]
I reviewed my work here again and realized that these parts of the big `if` clause:

```py
and not (old_field.remote_field and old_field.db_constraint)
and not (new_field.remote_field and new_field.db_constraint)
```

…were only there to prevent:

```
  File "/.../django/db/backends/sqlite3/schema.py", line 400, in _alter_field
    return super()._alter_field(
...
django.db.utils.OperationalError: no such constraint: "fk_0"
```

…which is actually the same `ALTER TABLE ... ALTER COLUMN ... DROP CONSTRAINT` bug that [is blocking the sister PR](https://github.com/django/django/pull/21181#discussion_r3145108561).

So I've dropped those parts, added a failing test, and will now mark this PR as draft until SQLite 3.53.1 is out.

**Comment by adamchainz** [comment]
> So I think this PR should keep constrained FK fields on `_remake_table()`, and leave FK fast-path support for a follow-up once both dropping and re-adding those constraints are supported.

Yeah, you're right, I was wrong. Restored those parts.


> (Additionally, [SQLite 3.53.1](https://sqlite.org/releaselog/3_53_1.html) has now been released.)

Yup, tested against it now, all green!
